internal/fs/parallel_dirops_test.go

```diff
@@ -0,0 +1,687 @@
+}
+
+func (t *ParallelDiropsTest) TestParallelLookUpsForSameDir() {
+	lookUpFunc := func(wg *sync.WaitGroup, dirPath string) (os.FileInfo, error) {
+		defer wg.Done()
+		fileInfo, err := os.Stat(dirPath)
+		return fileInfo, err
+	}
+	var stat1, stat2 os.FileInfo
+	var err1, err2 error
+
+	// Parallel lookups of explicit dir under mount.
+	dirPath := path.Join(mntDir, "explicitDir1")
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		stat1, err1 = lookUpFunc(&wg, dirPath)
+	}()
+	go func() {
+		stat2, err2 = lookUpFunc(&wg, dirPath)
+	}()
+	wg.Wait()
+
+	// Assert both stats passed and give correct information
+	assert.NoError(t.T(), err1)
+	assert.NoError(t.T(), err2)
+	assert.True(t.T(), stat1.IsDir())
+	assert.True(t.T(), stat2.IsDir())
+	assert.Contains(t.T(), dirPath, stat1.Name())
+	assert.Contains(t.T(), dirPath, stat2.Name())
+
+	// Parallel lookups of implicit dir in mount.
+	dirPath = path.Join(mntDir, "implicitDir1/")
+	wg.Add(2)
+	go func() {
+		stat1, err1 = lookUpFunc(&wg, dirPath)
+	}()
+	go func() {
+		stat2, err2 = lookUpFunc(&wg, dirPath)
+	}()
+	wg.Wait()
+
+	// Assert both stats passed and give correct information
+	assert.NoError(t.T(), err1)
+	assert.NoError(t.T(), err2)
+	assert.True(t.T(), stat1.IsDir())
+	assert.True(t.T(), stat2.IsDir())
+	assert.Contains(t.T(), dirPath, stat1.Name())
+	assert.Contains(t.T(), dirPath, stat2.Name())
+}
+
+func (t *ParallelDiropsTest) TestParallelReadDirsForSameDir() {
+	readDirFunc := func(wg *sync.WaitGroup, dirPath string) ([]os.DirEntry, error) {
+		defer wg.Done()
+		dirEntries, err := os.ReadDir(dirPath)
+		return dirEntries, err
+	}
+	var dirEntries1, dirEntries2 []os.DirEntry
+	var err1, err2 error
+
+	// Parallel readDirs of explicit dir under mount.
+	dirPath := path.Join(mntDir, "explicitDir1")
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		dirEntries1, err1 = readDirFunc(&wg, dirPath)
+	}()
+	go func() {
+		dirEntries2, err2 = readDirFunc(&wg, dirPath)
+	}()
+	wg.Wait()
+
+	// Assert both readDirs passed and give correct information
+	assert.NoError(t.T(), err1)
+	assert.NoError(t.T(), err2)
+	assert.Contains(t.T(), dirEntries1[0].Name(), "file1.txt")
+	assert.Contains(t.T(), dirEntries1[1].Name(), "file2.txt")
+	assert.Contains(t.T(), dirEntries2[0].Name(), "file1.txt")
+	assert.Contains(t.T(), dirEntries2[1].Name(), "file2.txt")
+
+	// Parallel readDirs of implicit dir under mount.
+	dirPath = path.Join(mntDir, "implicitDir1")
+	wg = sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		dirEntries1, err1 = readDirFunc(&wg, dirPath)
+	}()
+	go func() {
+		dirEntries2, err2 = readDirFunc(&wg, dirPath)
+	}()
+	wg.Wait()
+
+	// Assert both readDirs passed and give correct information
+	assert.NoError(t.T(), err1)
+	assert.NoError(t.T(), err2)
+	assert.Contains(t.T(), dirEntries1[0].Name(), "file1.txt")
+	assert.Contains(t.T(), dirEntries2[0].Name(), "file1.txt")
+
+	// Parallel readDirs of a directory and its parent directory.
+	dirPath = path.Join(mntDir, "explicitDir1")
```

Comment:
This test is contradicting the name of the test. Either we can change the name or move the test to another test.